
<!DOCTYPE html>



  


<html class="theme-next mist use-motion" lang="zh-Hans">
<head>
  <meta charset="UTF-8"/>
<meta http-equiv="X-UA-Compatible" content="IE=edge" />
<meta name="viewport" content="width=device-width, initial-scale=1, maximum-scale=1"/>
<meta name="theme-color" content="#222">









<meta http-equiv="Cache-Control" content="no-transform" />
<meta http-equiv="Cache-Control" content="no-siteapp" />



  <meta name="google-site-verification" content="9a0Rz8pHARxUTpjCCXlhEA66rXjiTW7jyoxyA6fM1N4" />













  
  
  <link href="/lib/fancybox/source/jquery.fancybox.css?v=2.1.5" rel="stylesheet" type="text/css" />




  
  
  
  

  
    
    
  

  

  

  

  

  
    
    
    <link href="//fonts.googleapis.com/css?family=Lato:300,300italic,400,400italic,700,700italic&subset=latin,latin-ext" rel="stylesheet" type="text/css">
  






<link href="/lib/font-awesome/css/font-awesome.min.css?v=4.6.2" rel="stylesheet" type="text/css" />

<link href="/css/main.css?v=5.1.2" rel="stylesheet" type="text/css" />


  <meta name="keywords" content="rnaseq,strand specific" />





  <link rel="alternate" href="/atom.xml" title="思考问题的熊" type="application/atom+xml" />




  <link rel="shortcut icon" type="image/x-icon" href="/favicon.ico?v=5.1.2" />






<meta name="description" content="关于链特异性测序的若干问题，很久以前就以为自己想清楚了，但是每次提起它的时候又容易重新产生各种各样的小困惑。于是整理一下，以免以后再时不时犯迷糊。很多东西就是这样，你以为的明白并不是真的明白，一年前的明白和一年后的明白也不是同一个明白。我这么说，不知道你能明白还是不明白。

 RNA-seq基本流程
下图是一个大概的RNA-seq基本流程

把RNA破碎成小片段，然后将RNA转变成一条cDNA，">
<meta property="og:type" content="article">
<meta property="og:title" content="链特异性测序那点事">
<meta property="og:url" content="http://kaopubear.top/2017-10-23-ssrnaseqbasic.html">
<meta property="og:site_name" content="思考问题的熊">
<meta property="og:description" content="关于链特异性测序的若干问题，很久以前就以为自己想清楚了，但是每次提起它的时候又容易重新产生各种各样的小困惑。于是整理一下，以免以后再时不时犯迷糊。很多东西就是这样，你以为的明白并不是真的明白，一年前的明白和一年后的明白也不是同一个明白。我这么说，不知道你能明白还是不明白。

 RNA-seq基本流程
下图是一个大概的RNA-seq基本流程

把RNA破碎成小片段，然后将RNA转变成一条cDNA，">
<meta property="og:image" content="http://oxzxnm82k.bkt.clouddn.com/17-10-22/39017681.jpg">
<meta property="og:image" content="http://oxzxnm82k.bkt.clouddn.com/17-10-23/32754319.jpg">
<meta property="og:image" content="http://oxzxnm82k.bkt.clouddn.com/17-10-23/3207114.jpg">
<meta property="og:image" content="http://oxzxnm82k.bkt.clouddn.com/17-10-22/61769932.jpg">
<meta property="og:image" content="http://oxzxnm82k.bkt.clouddn.com/17-10-22/62392221.jpg">
<meta property="og:image" content="http://oxzxnm82k.bkt.clouddn.com/17-10-22/80458612.jpg">
<meta property="og:image" content="http://oxzxnm82k.bkt.clouddn.com/17-10-22/61586179.jpg">
<meta property="og:image" content="http://oxzxnm82k.bkt.clouddn.com/17-10-23/828414.jpg">
<meta property="og:image" content="http://oxzxnm82k.bkt.clouddn.com/17-10-22/9379450.jpg">
<meta property="og:image" content="http://oxzxnm82k.bkt.clouddn.com/17-10-22/30847981.jpg">
<meta property="og:image" content="http://oxzxnm82k.bkt.clouddn.com/17-10-22/7418890.jpg">
<meta property="og:image" content="http://oxzxnm82k.bkt.clouddn.com/17-10-23/68849195.jpg">
<meta property="og:image" content="http://oxzxnm82k.bkt.clouddn.com/17-10-23/67993027.jpg">
<meta property="og:updated_time" content="2017-10-24T01:35:45.225Z">
<meta name="twitter:card" content="summary">
<meta name="twitter:title" content="链特异性测序那点事">
<meta name="twitter:description" content="关于链特异性测序的若干问题，很久以前就以为自己想清楚了，但是每次提起它的时候又容易重新产生各种各样的小困惑。于是整理一下，以免以后再时不时犯迷糊。很多东西就是这样，你以为的明白并不是真的明白，一年前的明白和一年后的明白也不是同一个明白。我这么说，不知道你能明白还是不明白。

 RNA-seq基本流程
下图是一个大概的RNA-seq基本流程

把RNA破碎成小片段，然后将RNA转变成一条cDNA，">
<meta name="twitter:image" content="http://oxzxnm82k.bkt.clouddn.com/17-10-22/39017681.jpg">



<script type="text/javascript" id="hexo.configurations">
  var NexT = window.NexT || {};
  var CONFIG = {
    root: '/',
    scheme: 'Mist',
    sidebar: {"position":"left","display":"post","offset":12,"offset_float":12,"b2t":false,"scrollpercent":false,"onmobile":false},
    fancybox: true,
    tabs: true,
    motion: true,
    duoshuo: {
      userId: 'undefined',
      author: '博主'
    },
    algolia: {
      applicationID: '',
      apiKey: '',
      indexName: '',
      hits: {"per_page":10},
      labels: {"input_placeholder":"Search for Posts","hits_empty":"We didn't find any results for the search: ${query}","hits_stats":"${hits} results found in ${time} ms"}
    }
  };
</script>



  <link rel="canonical" href="http://kaopubear.top/2017-10-23-ssrnaseqbasic.html"/>





  <title>链特异性测序那点事 | 思考问题的熊</title>
  




<script>
  (function(i,s,o,g,r,a,m){i['GoogleAnalyticsObject']=r;i[r]=i[r]||function(){
            (i[r].q=i[r].q||[]).push(arguments)},i[r].l=1*new Date();a=s.createElement(o),
          m=s.getElementsByTagName(o)[0];a.async=1;a.src=g;m.parentNode.insertBefore(a,m)
  })(window,document,'script','https://www.google-analytics.com/analytics.js','ga');
  ga('create', 'UA-91022805-1', 'auto');
  ga('send', 'pageview');
</script>


  <script type="text/javascript">
    var _hmt = _hmt || [];
    (function() {
      var hm = document.createElement("script");
      hm.src = "https://hm.baidu.com/hm.js?641a099d3da13af1d754c9098a3951fe";
      var s = document.getElementsByTagName("script")[0];
      s.parentNode.insertBefore(hm, s);
    })();
  </script>










</head>

<body itemscope itemtype="http://schema.org/WebPage" lang="zh-Hans">

  
  
    
  

  <div class="container sidebar-position-left page-post-detail ">
    <div class="headband"></div>

    <header id="header" class="header" itemscope itemtype="http://schema.org/WPHeader">
      <div class="header-inner"><div class="site-brand-wrapper">
  <div class="site-meta ">
    

    <div class="custom-logo-site-title">
      <a href="/"  class="brand" rel="start">
        <span class="logo-line-before"><i></i></span>
        <span class="site-title">思考问题的熊</span>
        <span class="logo-line-after"><i></i></span>
      </a>
    </div>
      
        <p class="site-subtitle">Data Thinker</p>
      
  </div>

  <div class="site-nav-toggle">
    <button>
      <span class="btn-bar"></span>
      <span class="btn-bar"></span>
      <span class="btn-bar"></span>
    </button>
  </div>
</div>

<nav class="site-nav">
  

  
    <ul id="menu" class="menu">
      
        
        <li class="menu-item menu-item-home">
          <a href="/" rel="section">
            
              <i class="menu-item-icon fa fa-fw fa-home"></i> <br />
            
            首页
          </a>
        </li>
      
        
        <li class="menu-item menu-item-about">
          <a href="/about" rel="section">
            
              <i class="menu-item-icon fa fa-fw fa-user"></i> <br />
            
            关于
          </a>
        </li>
      
        
        <li class="menu-item menu-item-categories">
          <a href="/categories" rel="section">
            
              <i class="menu-item-icon fa fa-fw fa-th"></i> <br />
            
            分类
          </a>
        </li>
      
        
        <li class="menu-item menu-item-tags">
          <a href="/tags" rel="section">
            
              <i class="menu-item-icon fa fa-fw fa-tags"></i> <br />
            
            标签
          </a>
        </li>
      
        
        <li class="menu-item menu-item-archives">
          <a href="/archives" rel="section">
            
              <i class="menu-item-icon fa fa-fw fa-archive"></i> <br />
            
            归档
          </a>
        </li>
      

      
        <li class="menu-item menu-item-search">
          
            <a href="javascript:;" class="popup-trigger">
          
            
              <i class="menu-item-icon fa fa-search fa-fw"></i> <br />
            
            搜索
          </a>
        </li>
      
    </ul>
  

  
    <div class="site-search">
      
  <div class="popup search-popup local-search-popup">
  <div class="local-search-header clearfix">
    <span class="search-icon">
      <i class="fa fa-search"></i>
    </span>
    <span class="popup-btn-close">
      <i class="fa fa-times-circle"></i>
    </span>
    <div class="local-search-input-wrapper">
      <input autocomplete="off"
             placeholder="搜索..." spellcheck="false"
             type="text" id="local-search-input">
    </div>
  </div>
  <div id="local-search-result"></div>
</div>



    </div>
  
</nav>

<link href="https://cdn.bootcss.com/KaTeX/0.7.1/katex.min.css" rel="stylesheet">
<script async src="//pagead2.googlesyndication.com/pagead/js/adsbygoogle.js"></script>
<script>
  (adsbygoogle = window.adsbygoogle || []).push({
    google_ad_client: "ca-pub-5362684605935255",
    enable_page_level_ads: true
  });
</script>

<script type="text/javascript" src="http://tajs.qq.com/stats?sId=63323019" charset="UTF-8"></script>

<script>
var _hmt = _hmt || [];
(function() {
  var hm = document.createElement("script");
  hm.src = "https://hm.baidu.com/hm.js?641a099d3da13af1d754c9098a3951fe";
  var s = document.getElementsByTagName("script")[0];
  s.parentNode.insertBefore(hm, s);
})();
</script>


 </div>
    </header>

    <main id="main" class="main">
      <div class="main-inner">
        <div class="content-wrap">
          <div id="content" class="content">
            

  <div id="posts" class="posts-expand">
    

  

  
  
  

  <article class="post post-type-normal" itemscope itemtype="http://schema.org/Article">
  
  
  
  <div class="post-block">
    <link itemprop="mainEntityOfPage" href="http://kaopubear.top/2017-10-23-ssrnaseqbasic.html">

    <span hidden itemprop="author" itemscope itemtype="http://schema.org/Person">
      <meta itemprop="name" content="Zhao Fei">
      <meta itemprop="description" content="">
      <meta itemprop="image" content="/uploads/myavatar.jpeg">
    </span>

    <span hidden itemprop="publisher" itemscope itemtype="http://schema.org/Organization">
      <meta itemprop="name" content="思考问题的熊">
    </span>

    
      <header class="post-header">

        
        
          <h1 class="post-title" itemprop="name headline">链特异性测序那点事</h1>
        

        <div class="post-meta">
          <span class="post-time">
            
              <span class="post-meta-item-icon">
                <i class="fa fa-calendar-o"></i>
              </span>
              
                <span class="post-meta-item-text">发表于</span>
              
              <time title="创建于" itemprop="dateCreated datePublished" datetime="2017-10-23T14:40:20+08:00">
                2017-10-23
              </time>
            

            

            
            
            <span class="post-letters-count">
              &nbsp; | &nbsp;
              <i class="fa fa-file-text-o" aria-hidden="true"></i>
              <span>4,090 字</span>
              &nbsp; | &nbsp;
              <i class="fa fa-clock-o" aria-hidden="true"></i>
              <span>约 17 min</span>
            </span>
            
          </span>

          
            <span class="post-category" >
            
              <span class="post-meta-divider">|</span>
            
              <span class="post-meta-item-icon">
                <i class="fa fa-folder-o"></i>
              </span>
              
                <span class="post-meta-item-text">分类于</span>
              
              
                <span itemprop="about" itemscope itemtype="http://schema.org/Thing">
                  <a href="/categories/生物信息/" itemprop="url" rel="index">
                    <span itemprop="name">生物信息</span>
                  </a>
                </span>

                
                
              
            </span>
          

          
            
            <!--noindex-->
              <span class="post-comments-count">
                <span class="post-meta-divider">|</span>
                <span class="post-meta-item-icon">
                  <i class="fa fa-comment-o"></i>
                </span>
                <a href="/2017-10-23-ssrnaseqbasic.html#comments" itemprop="discussionUrl">
                  <span class="post-comments-count hc-comment-count" data-xid="2017-10-23-ssrnaseqbasic.html" itemprop="commentsCount"></span>
                </a>
              </span>
              <!--/noindex-->
            
          

          
          
             <span id="/2017-10-23-ssrnaseqbasic.html" class="leancloud_visitors" data-flag-title="链特异性测序那点事">
               <span class="post-meta-divider">|</span>
               <span class="post-meta-item-icon">
                 <i class="fa fa-eye"></i>
               </span>
               
                 <span class="post-meta-item-text">阅读次数 </span>
               
                 <span class="leancloud-visitors-count"></span>
             </span>
          

          

          

          

        </div>
      </header>
    

    
    
    
    <div class="post-body" itemprop="articleBody">

      
      

      
        <blockquote>
<p>关于链特异性测序的若干问题，很久以前就以为自己想清楚了，但是每次提起它的时候又容易重新产生各种各样的小困惑。于是整理一下，以免以后再时不时犯迷糊。很多东西就是这样，你以为的明白并不是真的明白，一年前的明白和一年后的明白也不是同一个明白。我这么说，不知道你能明白还是不明白。</p>
</blockquote>
<h2 id="rna-seq基本流程"><a class="markdownIt-Anchor" href="#rna-seq基本流程"></a> RNA-seq基本流程</h2>
<p>下图是一个大概的RNA-seq基本流程</p>
<p><img src="http://oxzxnm82k.bkt.clouddn.com/17-10-22/39017681.jpg" alt=""></p>
<p>把RNA破碎成小片段，然后将RNA转变成一条cDNA，这一步需要用到反转录酶 reverse transcriptase (RT) 才能用RNA作为模板合成DNA。</p>
<p>不论是转录还是反转录都需要引物。通常如果我们要mRNA，那就可以用oligo-dT作为RT的引物，但是用它有两个问题，第一个是只能反转录那些有A尾巴的RNA，第二个问题是RT不是一个高度持续性的聚合酶，可能让转录提前发生终止，造成的结果就是3’端要比5’端reads富集，这样就会使得后续定量分析带来bias。</p>
<a id="more"></a>
<p>另一种常用的引物称为<strong>随机引物</strong>，随机引物的好处是没有A尾巴的诸如ncRNA也被留下了，而且不会存在明显的3’端偏差。但是很多研究也发现，所谓的随机引物根本就不随机，<strong>这也是测序结果中，通常前6个碱基的GC含量分布特别不均匀的原因</strong>。这几个碱基GC含量均匀很可能不是接头或者barcode那些东西，其实是Illumina 测序RT这一步的random hexamer priming 造成的bias，很多人在处理数据的时候会把这几个碱基去掉，其实很多时候真多RNA-seq数据去不去掉基本什么影响，不过开头如果有低质量的碱基倒是应该去掉。</p>
<p>随后是第二条链合成，这一步用是DNA聚合酶，以刚才和成的第一条链作为模板。</p>
<p>接下来就是在序列两端加上接头，加接头一方面是为了让机器可以识别这些序列，把这些序列固定；二是为了让多个样品可以同时上机，平摊每个样品的测序价格。双端测序为了让read从两边开始延伸，也需要在两端有所需的引物。</p>
<table>
<thead>
<tr>
<th>Adapter element</th>
<th>Requirement</th>
<th>Location</th>
<th>Function</th>
</tr>
</thead>
<tbody>
<tr>
<td><strong>Amplification element</strong></td>
<td>Required</td>
<td>5′ and 3′ terminus</td>
<td>Clonal amplification of the construct</td>
</tr>
<tr>
<td><strong>Primary sequencing priming site</strong></td>
<td>Required</td>
<td>Adjacent to the insert</td>
<td>Initiating the primary sequencing reaction</td>
</tr>
<tr>
<td><strong>Barcode/Index</strong></td>
<td>Optional</td>
<td>5′-end of the insert/Between the sequencing priming site and its respective amplification element</td>
<td>Provides a unique label to sequences from different samples. Allows pooling of multiple experiments in a single sequencing reaction.</td>
</tr>
<tr>
<td><strong>Paired-end sequencing priming site</strong></td>
<td>Optional</td>
<td>Adjacent to the insert on the side opposite of the primary sequencing priming site</td>
<td>Sequencing into the insert on the end opposite of the primary read</td>
</tr>
<tr>
<td><strong>Index sequencing priming site</strong></td>
<td>Optional</td>
<td>Complementary to the 5′-end of the sequencing priming site</td>
<td>Sequencing of the index</td>
</tr>
</tbody>
</table>
<p>所谓双端测序，因为很多时候read的长度要短于insert，为了增加覆盖度于是就想出了从insert两端同时测序的办法。使得测序深度增加的同时也能够用来判断isoform方向。</p>
<p>对于illumina数据，有一条5-3的universal adaptor；还有一条是3-5的indexed adatpor，这条引物含有特意的barcode。需要说明的是，在双端测序中，如果insert 不是足够长，那么R1可能就会测到R2的引物，同时R2可能会测到R1引物的反向互补序列。</p>
<p>大概的意思就是下面两张图。</p>
<p><img src="http://oxzxnm82k.bkt.clouddn.com/17-10-23/32754319.jpg" alt=""></p>
<p>加了接头以后进行PCR的扩增。扩增后就开始测序，测序的过程如下图所示。</p>
<p><img src="http://oxzxnm82k.bkt.clouddn.com/17-10-23/3207114.jpg" alt=""></p>
<p><img src="http://oxzxnm82k.bkt.clouddn.com/17-10-22/61769932.jpg" alt=""></p>
<p>测序的基本思想是机器识别四种碱基发出的不同颜色的荧光，可以理解为一个flow cell 立着非常多序列，机器一层一层扫过去，通过识别荧光而判断这一层每个序列的碱基是什么。</p>
<p>因为一个cell密密麻麻的全是荧光信号，机器并不是总能把每一个判断的非常准确，如果某一个荧光信号没有那么清晰，这个碱基的测序质量就比较低，如下图。</p>
<p><img src="http://oxzxnm82k.bkt.clouddn.com/17-10-22/62392221.jpg" alt=""></p>
<p>有的时候，如果一大片点都是同一种荧光，机器也可能犯晕，不知道到底哪一个荧光属于哪一个序列。这种情况尤其是在序列的前几个碱基容易发生。</p>
<blockquote>
<p>The sequencing machine uses the first few bases to establish where the cDNA fragments are on the flow cell. If all of the bases in one part of the flow cell are all the same, like ‘C’, and all show up green in the picture, then the colors will bleed together and it will not be clear where exactly all of the reads are. In contrast, if you have a lot of different colors in a region, it’s easier to determine where each one is, even with a little color bleed.</p>
</blockquote>
<p><img src="http://oxzxnm82k.bkt.clouddn.com/17-10-22/80458612.jpg" alt=""></p>
<h2 id="链特异性测序"><a class="markdownIt-Anchor" href="#链特异性测序"></a> 链特异性测序</h2>
<p>和普通的RNAseq不同，链特异性测序可以保留最初产生RNA的方向，普通建库方式为什么不行呢？因为传统建库方式通过两个接头的ligation把RNA已经变成了双链DNA，最后的文库中一部被测序的链对应正义链（sense strand），一部分被测序的链测是反义链。</p>
<p>链特异性建库方式有不止一种，对应到不同的软件又有不同的叫法，下面是几种称呼。<strong>要记住的是dUTP 测序方式的名字是fr-firstrand，也是RF。</strong> 至于具体的read方向接下来通过更详细的IGV截图说明问题。</p>
<p><img src="http://oxzxnm82k.bkt.clouddn.com/17-10-22/61586179.jpg" alt=""></p>
<p><img src="http://oxzxnm82k.bkt.clouddn.com/17-10-23/828414.jpg" alt=""></p>
<p>链特异性建库方式（以目前最常用的dUTP为例，如下图所示）首先利用随机引物合成RNA的一条cDNA链，在合成第二条链的时候用dUTP代替dTTP，加adaptor后用UDGase处理，将有U的第二条cDNA降解掉。</p>
<p><img src="http://oxzxnm82k.bkt.clouddn.com/17-10-22/9379450.jpg" alt=""></p>
<p>这样<strong>最后的insert DNA fragment都是来自于第一条cDNA，也就是dUTP叫fr-firststrand的原因</strong>。对于dUTP数据，<strong>tophat</strong>的参数应该为<code>–library-type fr-firststrand</code>。这里的first-strand cDNA可不是RNA strand，在使用<strong>htseq-count</strong> 时，真正的正义链应该是使用参数<code>-s reverse</code> 得到的结果。</p>
<h2 id="正正反反不清楚"><a class="markdownIt-Anchor" href="#正正反反不清楚"></a> 正正反反不清楚</h2>
<p>说到链特异性测序，实在让人困惑的是<strong>各种链的概念</strong>，尤其是翻译成中文就更说不清了。</p>
<p>DNA 的正链和负链，就是那两条反向互补的链。参考基因组给出的那个链就是所谓的正链（forword），另一条链是反链（reverse）。但是这正反一定<strong>不能和正义链（sense strand）反义链（antisense strand）混淆</strong>，两条互补的DNA链其中一条携带编码蛋白质信息的链称为正义链，另一条与之互补的称为反义链。但是携带编码信息的正义链不是模板，只是因为它的序列和RNA相同，正义链也是编码链。而反义链虽然和RNA反向互补，但它可是真正给RNA当模板的链，因此反义链也是模板链。</p>
<p><strong>总结两点</strong></p>
<ol>
<li>
<p><strong>正义链（sense strand）= 编码链（coding strand）= 非模板链</strong></p>
</li>
<li>
<p><strong>forword strand 上可以同时有sense strand 和 antisense strand。因为这完全是两个不同的概念。</strong></p>
</li>
</ol>
<p>写这篇文章的原因，就是因为有人问我，链特异性测序数据 htseq-count 的结果是不是应该把正负链的基因分别在-s yes 和-s reverse 两个参数结果中统计出来再做下游分析。这里犯的错误就是我们混淆了基因组正反链和基因正义反义链的概念。</p>
<h2 id="dutp到底是怎么回事"><a class="markdownIt-Anchor" href="#dutp到底是怎么回事"></a> dUTP到底是怎么回事</h2>
<p>从前文的一个图我们可以总结出<strong>dUTP方式测序R1文件中read1 的方向和基因的方向（正义链）是相反的，而R2文件中的read2 方向和基因的方向是相同的。</strong></p>
<p>可以参考下面的两个igv文件bam截图。</p>
<p><strong>首先解释一下igv 两个颜色参数的意义</strong></p>
<ul>
<li><strong>Read strand</strong> in pastels, <strong>red for positive rightward (5’ to 3’) DNA strand</strong>, <strong>blue for negative leftward (reverse-complement) DNA strand</strong>, and grey for unpaired mate, mate not mapped, or otherwise unknown status.</li>
<li><strong>First-of-pair strand</strong> assignment is dependent on RNA transcript directionality and is <strong>useful for directional libraries.</strong> <strong>Displays reads or read pairs in which the forward read is first (F1 or F1R2) in red and reads or read pairs in which the reverse read is first (R1 or R1F2) in blue.</strong> Unknown status is in gray.
<ul>
<li>For a given transcript, non-directional libraries will show a mix of red and blue reads aligning to the locus.</li>
<li><strong>Directional libraries will show reads of one color in the direction matching the transcript orientation.</strong></li>
</ul>
</li>
</ul>
<p>下面这个图示按照igv 颜色选项中的read strand 方向进行区分，可以看到所有<strong>红色read都是在正链方向</strong>（注意正链不是正义链），而所有<strong>蓝色的read都是负链方向</strong>。下面基因的方向是正链方向，也就是和粉色的read同向的，如果你把鼠标放到随意一个粉色的read上，就能看到显示的信息是<strong>second of pair</strong>，也就是<strong>pair中的read2（R2）</strong>；反之如果你在蓝色的read上面，就会显示信息是first of pair，也就是R1 。</p>
<p><strong>总结，dUTP测序中pair read 中的read1（R1）和基因方向相反，read2（R2）和基因方向相同</strong></p>
<p><img src="http://oxzxnm82k.bkt.clouddn.com/17-10-22/30847981.jpg" alt=""></p>
<p>再看下面这张图</p>
<p><img src="http://oxzxnm82k.bkt.clouddn.com/17-10-22/7418890.jpg" alt=""></p>
<p>这张图展示了两个基因1和2，我们可以发现<strong>gene1的正义链就在正链上</strong>，而<strong>gene2的正义链其实是在反链上</strong>。看read情况，a，c两个read虽然针对正链负链而言方向一致，都是负链方向，但是如果把<strong>a是pair中的read1（first of pair ）</strong>，而<strong>c是pair中的read2（second of pair）</strong>。也就是说，<strong>read方向一致，但一个是read1一个是read2，说明这两个read对应的基因一定是反向的。<strong>同样的道理，虽然</strong>b，d都是两个方向为负链的read，但是b其实是所在pair的read2（second of pair），而d是所在pair的read1（first of pair）。</strong></p>
<p><strong>再次强调，dUTP测序中pair read 中的read1（R1）和基因方向相反，read2和基因方向相同</strong></p>
<p>当使用read strand来进行颜色区分的时候，每一个基因上两种颜色的分布应该相对均匀（也就是所谓的pair end）。</p>
<p>如果这个时候把<strong>颜色选项改为按照<code>first of pair of strand</code>来区分</strong>，会出现下图的变化。</p>
<p><img src="http://oxzxnm82k.bkt.clouddn.com/17-10-23/68849195.jpg" alt=""></p>
<p>geng1的read全部变成了紫色，而gene2的read全部变成了粉色。</p>
<p>如果是非链特异性测序，在<code>first of pair of strand</code>模式下，同一个gene相关的read颜色也是明显混杂的。如下图</p>
<p><img src="http://oxzxnm82k.bkt.clouddn.com/17-10-23/67993027.jpg" alt=""></p>
<p><strong>再一次总结：</strong></p>
<ul>
<li>dUTP 链特异性测序中，RNA 方向（gff文件中基因的方向）与read1相反，与read2相同。如果read1比对到基因组正链上，则对应的gene在基因组负链；如果read2比对到基因组正链则对应的gene在基因组正链。</li>
<li>dTUP 测序方式叫做fr-firstrand（留下的是cDNA第一条链），也是RF。</li>
<li>如果dUTP链特异性测序，看基因表达量应该 counts for the 2nd read strand aligned with RNA(htseq-count option -s reverse, STAR ReadsPerGene.out.tab column 3 )</li>
<li>如果想看反义链是否有转录本（比如NAT）应该用 the 1st read strand aligned with RNA ( htseq-count option -s yes，STAR ReadsPerGene.out.tab column 4)</li>
</ul>
<h2 id="几个常用软件的设置"><a class="markdownIt-Anchor" href="#几个常用软件的设置"></a> 几个常用软件的设置</h2>
<p><strong>STAR</strong> mpping 时无需特别设置，但如果不是链特异性数据且下游分析要用到cufflinks 则需要增加参数 --outSAMstrandField intronMotif。为的是增加一个XS标签。</p>
<blockquote>
<p>If you have <strong>stranded RNA-seq data, you do not need to use any specific STAR options.</strong> Instead, you need to run Cufflinks with the library option --library-type options. For example, <strong>cufflinks… --library-type fr-firststrand</strong> should be used for the standard dUTP protocol, including Illumina’s stranded Tru-Seq.</p>
</blockquote>
<p><strong>hisat2 --rna-strandness RF</strong></p>
<p>目的也是给比对序列添加一个XS标签以区分方向，方面cufflinks使用。</p>
<blockquote>
<p>For single-end reads, use F or R. ‘F’ means a read corresponds to a transcript. ‘R’ means a read corresponds to the reverse complemented counterpart of a transcript. For paired-end reads, use either FR or RF.<br>
With this option being used, every read alignment will have an XS attribute tag: ‘+’ means a read belongs to a transcript on ‘+’ strand of genome. ‘-’ means a read belongs to a transcript on ‘-’ strand of genome.</p>
</blockquote>
<p><strong>tophat</strong>  <strong>–library-type option  fr-firststrand</strong></p>
<p>具体解释参见下表</p>
<table>
<thead>
<tr>
<th><strong>Library Type</strong></th>
<th><strong>Examples</strong></th>
<th><strong>Description</strong></th>
</tr>
</thead>
<tbody>
<tr>
<td>fr-unstranded</td>
<td>Standard Illumina</td>
<td>Reads from the left-most end of the fragment (in transcript coordinates) map to the transcript strand, and the right-most end maps to the opposite strand.</td>
</tr>
<tr>
<td>fr-firststrand</td>
<td>dUTP, NSR, NNSR</td>
<td>Same as above except we enforce the rule that the right-most end of the fragment (in transcript coordinates) is the first sequenced (or only sequenced for single-end reads). Equivalently, it is assumed that only the strand generated during first strand synthesis is sequenced.</td>
</tr>
<tr>
<td>fr-secondstrand</td>
<td>Ligation, Standard SOLiD</td>
<td>Same as above except we enforce the rule that the left-most end of the fragment (in transcript coordinates) is the first sequenced (or only sequenced for single-end reads). Equivalently, it is assumed that only the strand generated during second strand synthesis is sequenced.</td>
</tr>
</tbody>
</table>
<p><strong>htseq-count -s reverse/yes(看反义链)</strong></p>
<blockquote>
<p>For <code>stranded=no</code>, a read is considered overlapping with a feature regardless of whether it is mapped to the same or the opposite strand as the feature. For <code>stranded=yes</code> and single-end reads, the read has to be mapped to the same strand as the feature. For paired-end reads, the first read has to be on the same strand and the second read on the opposite strand. For <code>stranded=reverse</code>, these rules are reversed.</p>
</blockquote>
<p><strong>RSEM --forward-prob  0（正义链）1（看反义链）</strong></p>
<blockquote>
<p>The RNA-Seq protocol used to generate the reads is strand specific, i.e., <strong>all (upstream) reads are derived from the forward strand</strong>. <strong>This option is equivalent to --forward-prob=1.0</strong>. With this option set, if RSEM runs the Bowtie/Bowtie 2 aligner, the ‘–norc’ Bowtie/Bowtie 2 option will be used, which disables alignment to the reverse strand of transcripts. (Default: off)</p>
<p>Probability of generating a read from the forward strand of a transcript. Set to 1 for a strand-specific protocol where all (upstream) reads are derived from the forward strand, <strong>0 for a strand-specific protocol where all (upstream) read are derived from the reverse strand</strong>, or 0.5 for a non-strand-specific protocol. (Default: 0.5)</p>
</blockquote>
<p><strong>sXpress --rf-stranded / --fr-stranded(看反义链)</strong></p>
<blockquote>
<p>–fr  eXpress only accepts alignments (single-end or paired) where the first (or only) read is aligned to the forward target sequence and the second read is aligned to the reverse-complemented target sequence. In directional sequencing, this is equivalent to second-strand only.</p>
<p>–rf eXpress only accepts alignments (single-end or paired) where the first (or only) read is aligned to the reverse-completemented target sequence and the second read is aligned to the forward target sequence. In directional sequencing, this is equivalent to first-strand only.</p>
</blockquote>
<p><strong>trinity --SS_lib_type RF</strong></p>
<blockquote>
<p>Trinity performs best with strand-specific data, in which case sense and antisense transcripts can be resolved.</p>
<p><em>RF</em>: first read (/1) of fragment pair is sequenced as anti-sense (reverse(<em>R</em>)), and second read (/2) is in the sense strand (forward(<em>F</em>)); typical of the dUTP/UDG sequencing method.</p>
</blockquote>
<h2 id="参数错了又怎样"><a class="markdownIt-Anchor" href="#参数错了又怎样"></a> 参数错了又怎样？</h2>
<p>到这里，会想问两个问题。有时候我们不知道数据的建库方式是不是链特异性的，如果弄错了结果会怎么样呢？</p>
<p>如果你用STAR mapping 完可以用igv像上文提到的那样，看看是不是链特异性测序。</p>
<p>下面是两个真是数据的count 统计情况。</p>
<p>对于<strong>仅仅进行基因表达定量</strong>来说，把链特异性数据当作普通建库数据来处理，可以观察第2列数据和第4列数据。具体某一个基因而言，影响不会太大，因为绝大多数反义链本身表达量就非常低。<br>
不过可以注意 noFeature 和 ambiguous 这两个值，因为基因组中存在两个基因分别在正链和负链且又重叠的情况，不区分方向会比区分方向的ambiguous数目多一些。因为如果不能通过方向来区分到底属于哪个基因，这样的read就会被认为是ambiguous。</p>
<p>但是因为区分了方向，又会使得noFeature的数目更多一些。不过两者总体影响不会差别非常大。如果不能判断建库方式，在htseq中使用-s no 参数是一个比较保险（虽然不是非常精确）的做法。</p>
<table>
<thead>
<tr>
<th></th>
<th>-s no</th>
<th>-s yes</th>
<th>-s reverse</th>
</tr>
</thead>
<tbody>
<tr>
<td>N_noFeature</td>
<td>1290001</td>
<td>16837194</td>
<td>1480658</td>
</tr>
<tr>
<td>N_ambiguous</td>
<td>633021</td>
<td>16413</td>
<td>74710</td>
</tr>
<tr>
<td>AT1G01010</td>
<td>58</td>
<td>2</td>
<td>56</td>
</tr>
<tr>
<td>AT1G01020</td>
<td>65</td>
<td>0</td>
<td>65</td>
</tr>
<tr>
<td>AT1G03987</td>
<td>5</td>
<td>5</td>
<td>0</td>
</tr>
<tr>
<td>AT1G01030</td>
<td>296</td>
<td>5</td>
<td>291</td>
</tr>
<tr>
<td>AT1G01040</td>
<td>901</td>
<td>5</td>
<td>1078</td>
</tr>
<tr>
<td>AT1G03993</td>
<td>0</td>
<td>182</td>
<td>0</td>
</tr>
<tr>
<td>AT1G01050</td>
<td>428</td>
<td>0</td>
<td>434</td>
</tr>
<tr>
<td>AT1G03997</td>
<td>1</td>
<td>7</td>
<td>0</td>
</tr>
<tr>
<td>AT1G01060</td>
<td>85</td>
<td>0</td>
<td>85</td>
</tr>
<tr>
<td>AT1G01070</td>
<td>73</td>
<td>0</td>
<td>73</td>
</tr>
<tr>
<td>AT1G04003</td>
<td>0</td>
<td>0</td>
<td>0</td>
</tr>
<tr>
<td>AT1G01080</td>
<td>1166</td>
<td>15</td>
<td>1151</td>
</tr>
<tr>
<td>AT1G01090</td>
<td>2901</td>
<td>0</td>
<td>2901</td>
</tr>
<tr>
<td>AT1G01100</td>
<td>1560</td>
<td>0</td>
<td>1560</td>
</tr>
<tr>
<td>AT1G01110</td>
<td>82</td>
<td>0</td>
<td>82</td>
</tr>
<tr>
<td>AT1G01120</td>
<td>484</td>
<td>0</td>
<td>484</td>
</tr>
<tr>
<td>AT1G01130</td>
<td>72</td>
<td>9</td>
<td>63</td>
</tr>
<tr>
<td>AT1G01140</td>
<td>518</td>
<td>3</td>
<td>515</td>
</tr>
<tr>
<td>AT1G01150</td>
<td>0</td>
<td>1</td>
<td>0</td>
</tr>
<tr>
<td>AT1G01160</td>
<td>356</td>
<td>192</td>
<td>551</td>
</tr>
<tr>
<td>AT1G04007</td>
<td>4</td>
<td>189</td>
<td>0</td>
</tr>
<tr>
<td>AT1G01170</td>
<td>55</td>
<td>11</td>
<td>423</td>
</tr>
</tbody>
</table>
<p>相反，如果把普通建库方式的数据当作链特异性数据来处理。</p>
<p>比如在htseq-count中使用了-s reverse 参数，这个时候<strong>只有R2方向和基因方向相同的pair才用来算作一个count</strong>，所有R2和基因方向不同的pair就被当作no feature了。这样的结果影响可以通过下面的表格观察。</p>
<p>用正常方法数出的noFeature 是6万左右，而用-s yes或者reverse数出来的noFeature 就接近46万了。将近40万的read 被丢掉。</p>
<p>所以，如果把普通建库的数据误当作链特异性数据来处理极有可能会损失大量的数据，如果<strong>弄错了链特异性建库的方式</strong>，那坑你就没几个read剩下了。另外，计算出来的结果自然也会有非常大的差异，是不准确的。</p>
<table>
<thead>
<tr>
<th></th>
<th>-s no</th>
<th>-s yes</th>
<th>-s reverse</th>
</tr>
</thead>
<tbody>
<tr>
<td>N_unmapped</td>
<td>729831</td>
<td>729831</td>
<td>729831</td>
</tr>
<tr>
<td>N_multimapping</td>
<td>443861</td>
<td>443861</td>
<td>443861</td>
</tr>
<tr>
<td>N_noFeature</td>
<td>63787</td>
<td>4591673</td>
<td>4599916</td>
</tr>
<tr>
<td>N_ambiguous</td>
<td>594720</td>
<td>27992</td>
<td>27789</td>
</tr>
<tr>
<td>AT1G01010</td>
<td>101</td>
<td>54</td>
<td>47</td>
</tr>
<tr>
<td>AT1G01020</td>
<td>84</td>
<td>41</td>
<td>43</td>
</tr>
<tr>
<td>AT1G03987</td>
<td>1</td>
<td>0</td>
<td>1</td>
</tr>
<tr>
<td>AT1G01030</td>
<td>80</td>
<td>37</td>
<td>43</td>
</tr>
<tr>
<td>AT1G01040</td>
<td>499</td>
<td>279</td>
<td>293</td>
</tr>
<tr>
<td>AT1G03993</td>
<td>0</td>
<td>33</td>
<td>25</td>
</tr>
<tr>
<td>AT1G01050</td>
<td>634</td>
<td>312</td>
<td>337</td>
</tr>
<tr>
<td>AT1G03997</td>
<td>0</td>
<td>0</td>
<td>0</td>
</tr>
<tr>
<td>AT1G01060</td>
<td>3274</td>
<td>1644</td>
<td>1630</td>
</tr>
<tr>
<td>AT1G01070</td>
<td>97</td>
<td>43</td>
<td>54</td>
</tr>
<tr>
<td>AT1G04003</td>
<td>0</td>
<td>0</td>
<td>0</td>
</tr>
<tr>
<td>AT1G01080</td>
<td>585</td>
<td>303</td>
<td>282</td>
</tr>
<tr>
<td>AT1G01090</td>
<td>1768</td>
<td>907</td>
<td>861</td>
</tr>
<tr>
<td>AT1G01100</td>
<td>1132</td>
<td>549</td>
<td>583</td>
</tr>
<tr>
<td>AT1G01110</td>
<td>60</td>
<td>21</td>
<td>39</td>
</tr>
<tr>
<td>AT1G01120</td>
<td>1099</td>
<td>573</td>
<td>526</td>
</tr>
</tbody>
</table>
<p><a href="https://rnaseq.uoregon.edu/" target="_blank" rel="external">参考资料1</a></p>
<p><a href="http://onetipperday.sterding.com/2013/06/illumina-hiseq2000-adaptor.html" target="_blank" rel="external">参考资料2</a></p>
<p><a href="http://onetipperday.sterding.com/2013/12/illumina-hiseq2000-adaptor-and.html" target="_blank" rel="external">参考资料3</a></p>

      
    </div>
    <div>
      
      
        <ul class="post-copyright">
          <li class="post-copyright-author">
              <strong>作者：</strong>Zhao Fei
          </li>
          <li class="post-copyright-link">
            <strong>链接：</strong>
            <a href="/2017-10-23-ssrnaseqbasic.html" title="链特异性测序那点事">2017-10-23-ssrnaseqbasic.html</a>
          </li>
          <li class="post-copyright-link">
            <strong>说明：</strong>
            博客所有文章均仅代表个人观点，欢迎在评论区留言交流共同进步。
          </li>
          <li class="post-copyright-license">
            <strong>版权： </strong>
            博客所有文章除特别声明均采用 <a href="https://creativecommons.org/licenses/by-nc-sa/4.0/deed.zh" rel="external  nofollow" target="_blank">CC BY-NC-SA 4.0 CN</a>   许可协议。转载请<a href="http://kaopubear.top/about/" target="_blank">联系作者</a>并注明出处！
          </li>
        </ul>
      
    </div>
    
    
    

    

    
      <div>
        <div style="padding: 10px 0; margin: 20px auto; width: 90%; text-align: center;">
  <div>鼓励创作</div>
  <button id="rewardButton" disable="enable" onclick="var qr = document.getElementById('QR'); if (qr.style.display === 'none') {qr.style.display='block';} else {qr.style.display='none'}">
    <span>赞赏</span>
  </button>
  <div id="QR" style="display: none;">

    
      <div id="wechat" style="display: inline-block">
        <img id="wechat_qr" src="/uploads/wechat.jpg" alt="Zhao Fei WeChat Pay"/>
        <p>WeChat Pay</p>
      </div>
    

    
      <div id="alipay" style="display: inline-block">
        <img id="alipay_qr" src="/uploads/alipay.jpg" alt="Zhao Fei Alipay"/>
        <p>Alipay</p>
      </div>
    

    

  </div>
</div>

      </div>
    

    

    <footer class="post-footer">
      
        <div class="post-tags">
          
            <a href="/tags/bioinformatics/" rel="tag"># bioinformatics</a>
          
            <a href="/tags/RNA-seq/" rel="tag"># RNA-seq</a>
          
        </div>
      

      
      
      

      
        <div class="post-nav">
          <div class="post-nav-next post-nav-item">
            
              <a href="/2017-10-22-shellscript1.html" rel="next" title="shell脚本应知应会">
                <i class="fa fa-chevron-left"></i> shell脚本应知应会
              </a>
            
          </div>

          <span class="post-nav-divider"></span>

          <div class="post-nav-prev post-nav-item">
            
          </div>
        </div>
      

      
      
    </footer>
  </div>
  
  
  
  </article>



    <div class="post-spread">
      
        <!-- JiaThis Button BEGIN -->
<div class="jiathis_style">
  <a class="jiathis_button_tsina"></a>
  <a class="jiathis_button_tqq"></a>
  <a class="jiathis_button_weixin"></a>
  <a class="jiathis_button_cqq"></a>
  <a class="jiathis_button_douban"></a>
  <a class="jiathis_button_renren"></a>
  <a class="jiathis_button_qzone"></a>
  <a class="jiathis_button_kaixin001"></a>
  <a class="jiathis_button_copy"></a>
  <a href="http://www.jiathis.com/share" class="jiathis jiathis_txt jiathis_separator jtico jtico_jiathis" target="_blank"></a>
  <a class="jiathis_counter_style"></a>
</div>
<script type="text/javascript" >
  var jiathis_config={
    hideMore:false
  }
</script>
<script type="text/javascript" src="http://v3.jiathis.com/code/jia.js" charset="utf-8"></script>
<!-- JiaThis Button END -->

      
    </div>
  </div>


          </div>
          


          
  <div class="comments" id="comments">
    
      <div id="hypercomments_widget"></div>
    
  </div>


        </div>
        
          
  
  <div class="sidebar-toggle">
    <div class="sidebar-toggle-line-wrap">
      <span class="sidebar-toggle-line sidebar-toggle-line-first"></span>
      <span class="sidebar-toggle-line sidebar-toggle-line-middle"></span>
      <span class="sidebar-toggle-line sidebar-toggle-line-last"></span>
    </div>
  </div>

  <aside id="sidebar" class="sidebar">
    
    <div class="sidebar-inner">

      

      
        <ul class="sidebar-nav motion-element">
          <li class="sidebar-nav-toc sidebar-nav-active" data-target="post-toc-wrap" >
            文章目录
          </li>
          <li class="sidebar-nav-overview" data-target="site-overview">
            站点概览
          </li>
        </ul>
      

      <section class="site-overview sidebar-panel">
        <div class="site-author motion-element" itemprop="author" itemscope itemtype="http://schema.org/Person">
          <img class="site-author-image" itemprop="image"
               src="/uploads/myavatar.jpeg"
               alt="Zhao Fei" />
          <p class="site-author-name" itemprop="name">Zhao Fei</p>
           
              <p class="site-description motion-element" itemprop="description">Bioinformatics PhD student</br> 如果看到博客的你感觉在这里有一点收获，对我最好的奖赏就是请你也教我一点你认为我该知道的知识。</p>
          
        </div>
        <nav class="site-state motion-element">

          
            <div class="site-state-item site-state-posts">
              <a href="/archives">
                <span class="site-state-item-count">108</span>
                <span class="site-state-item-name">日志</span>
              </a>
            </div>
          

          
            
            
            <div class="site-state-item site-state-categories">
              <a href="/categories/index.html">
                <span class="site-state-item-count">7</span>
                <span class="site-state-item-name">分类</span>
              </a>
            </div>
          

          
            
            
            <div class="site-state-item site-state-tags">
              <a href="/tags/index.html">
                <span class="site-state-item-count">29</span>
                <span class="site-state-item-name">标签</span>
              </a>
            </div>
          

        </nav>

        
          <div class="feed-link motion-element">
            <a href="/atom.xml" rel="alternate">
              <i class="fa fa-rss"></i>
              RSS
            </a>
          </div>
        

        <div class="links-of-author motion-element">
          
            
              <span class="links-of-author-item">
                <a href="Mailto:ghsau@163.com" target="_blank" title="E-Mail">
                  
                    <i class="fa fa-fw fa-envelope"></i>
                  
                    
                      E-Mail
                    
                </a>
              </span>
            
              <span class="links-of-author-item">
                <a href="http://www.jianshu.com/u/963b2c9fee02" target="_blank" title="简书">
                  
                    <i class="fa fa-fw fa-pencil"></i>
                  
                    
                      简书
                    
                </a>
              </span>
            
              <span class="links-of-author-item">
                <a href="http://weibo.com/277440877" target="_blank" title="Weibo">
                  
                    <i class="fa fa-fw fa-weibo"></i>
                  
                    
                      Weibo
                    
                </a>
              </span>
            
              <span class="links-of-author-item">
                <a href="http://kaopubear.top/uploads/qrcode_for_ibear_share.jpg" target="_blank" title="Wechat">
                  
                    <i class="fa fa-fw fa-weixin"></i>
                  
                    
                      Wechat
                    
                </a>
              </span>
            
              <span class="links-of-author-item">
                <a href="http://book.kaopubear.top" target="_blank" title="GitBook">
                  
                    <i class="fa fa-fw fa-book"></i>
                  
                    
                      GitBook
                    
                </a>
              </span>
            
          
        </div>

        
        

        
        

        


      </section>

      
      <!--noindex-->
        <section class="post-toc-wrap motion-element sidebar-panel sidebar-panel-active">
          <div class="post-toc">

            
              
            

            
              <div class="post-toc-content"><ol class="nav"><li class="nav-item nav-level-2"><a class="nav-link" href="#rna-seq基本流程"><span class="nav-number">1.</span> <span class="nav-text"> RNA-seq基本流程</span></a></li><li class="nav-item nav-level-2"><a class="nav-link" href="#链特异性测序"><span class="nav-number">2.</span> <span class="nav-text"> 链特异性测序</span></a></li><li class="nav-item nav-level-2"><a class="nav-link" href="#正正反反不清楚"><span class="nav-number">3.</span> <span class="nav-text"> 正正反反不清楚</span></a></li><li class="nav-item nav-level-2"><a class="nav-link" href="#dutp到底是怎么回事"><span class="nav-number">4.</span> <span class="nav-text"> dUTP到底是怎么回事</span></a></li><li class="nav-item nav-level-2"><a class="nav-link" href="#几个常用软件的设置"><span class="nav-number">5.</span> <span class="nav-text"> 几个常用软件的设置</span></a></li><li class="nav-item nav-level-2"><a class="nav-link" href="#参数错了又怎样"><span class="nav-number">6.</span> <span class="nav-text"> 参数错了又怎样？</span></a></li></ol></div>
            

          </div>
        </section>
      <!--/noindex-->
      

      

    </div>
  </aside>


        
      </div>
    </main>

    <footer id="footer" class="footer">
      <div class="footer-inner">
        <div class="copyright" >
  
  &copy; 
  <span itemprop="copyrightYear">2017</span>
  <span class="with-love">
    <i class="fa fa-heart"></i>
  </span>
  <span class="author" itemprop="copyrightHolder">Zhao Fei</span>
  &nbsp; | &nbsp;
  <span class="post-meta-item-icon">
    <i class="fa fa-file-text" aria-hidden="true"></i>
  </span
  <span class="post-count">total words: 230.7k</span>
</div>



        
<div class="busuanzi-count">
  <script async src="https://dn-lbstatics.qbox.me/busuanzi/2.3/busuanzi.pure.mini.js"></script>

  
    <span class="site-uv">
      <i class="fa fa-user"></i>
      <span class="busuanzi-value" id="busuanzi_value_site_uv"></span>
      
    </span>
  

  
    <span class="site-pv">
      <i class="fa fa-eye"></i>
      <span class="busuanzi-value" id="busuanzi_value_site_pv"></span>
      
    </span>
  
</div>


        
      </div>
    </footer>

    
      <div class="back-to-top">
        <i class="fa fa-arrow-up"></i>
        
      </div>
    

  </div>

  

<script type="text/javascript">
  if (Object.prototype.toString.call(window.Promise) !== '[object Function]') {
    window.Promise = null;
  }
</script>









  












  
  <script type="text/javascript" src="/lib/jquery/index.js?v=2.1.3"></script>

  
  <script type="text/javascript" src="/lib/fastclick/lib/fastclick.min.js?v=1.0.6"></script>

  
  <script type="text/javascript" src="/lib/jquery_lazyload/jquery.lazyload.js?v=1.9.7"></script>

  
  <script type="text/javascript" src="/lib/velocity/velocity.min.js?v=1.2.1"></script>

  
  <script type="text/javascript" src="/lib/velocity/velocity.ui.min.js?v=1.2.1"></script>

  
  <script type="text/javascript" src="/lib/fancybox/source/jquery.fancybox.pack.js?v=2.1.5"></script>


  


  <script type="text/javascript" src="/js/src/utils.js?v=5.1.2"></script>

  <script type="text/javascript" src="/js/src/motion.js?v=5.1.2"></script>



  
  

  
  <script type="text/javascript" src="/js/src/scrollspy.js?v=5.1.2"></script>
<script type="text/javascript" src="/js/src/post-details.js?v=5.1.2"></script>



  


  <script type="text/javascript" src="/js/src/bootstrap.js?v=5.1.2"></script>



  


  




	

		<script type="text/javascript">
		_hcwp = window._hcwp || [];

		_hcwp.push({widget:"Bloggerstream", widget_id: 94267, selector:".hc-comment-count", label: "{\%COUNT%\}" });

		
		_hcwp.push({widget:"Stream", widget_id: 94267, xid: "2017-10-23-ssrnaseqbasic.html"});
		

		(function() {
		if("HC_LOAD_INIT" in window)return;
		HC_LOAD_INIT = true;
		var lang = (navigator.language || navigator.systemLanguage || navigator.userLanguage || "en").substr(0, 2).toLowerCase();
		var hcc = document.createElement("script"); hcc.type = "text/javascript"; hcc.async = true;
		hcc.src = ("https:" == document.location.protocol ? "https" : "http")+"://w.hypercomments.com/widget/hc/94267/"+lang+"/widget.js";
		var s = document.getElementsByTagName("script")[0];
		s.parentNode.insertBefore(hcc, s.nextSibling);
		})();
		</script>

	










  

  <script type="text/javascript">
    // Popup Window;
    var isfetched = false;
    var isXml = true;
    // Search DB path;
    var search_path = "search.xml";
    if (search_path.length === 0) {
      search_path = "search.xml";
    } else if (/json$/i.test(search_path)) {
      isXml = false;
    }
    var path = "/" + search_path;
    // monitor main search box;

    var onPopupClose = function (e) {
      $('.popup').hide();
      $('#local-search-input').val('');
      $('.search-result-list').remove();
      $('#no-result').remove();
      $(".local-search-pop-overlay").remove();
      $('body').css('overflow', '');
    }

    function proceedsearch() {
      $("body")
        .append('<div class="search-popup-overlay local-search-pop-overlay"></div>')
        .css('overflow', 'hidden');
      $('.search-popup-overlay').click(onPopupClose);
      $('.popup').toggle();
      var $localSearchInput = $('#local-search-input');
      $localSearchInput.attr("autocapitalize", "none");
      $localSearchInput.attr("autocorrect", "off");
      $localSearchInput.focus();
    }

    // search function;
    var searchFunc = function(path, search_id, content_id) {
      'use strict';

      // start loading animation
      $("body")
        .append('<div class="search-popup-overlay local-search-pop-overlay">' +
          '<div id="search-loading-icon">' +
          '<i class="fa fa-spinner fa-pulse fa-5x fa-fw"></i>' +
          '</div>' +
          '</div>')
        .css('overflow', 'hidden');
      $("#search-loading-icon").css('margin', '20% auto 0 auto').css('text-align', 'center');

      $.ajax({
        url: path,
        dataType: isXml ? "xml" : "json",
        async: true,
        success: function(res) {
          // get the contents from search data
          isfetched = true;
          $('.popup').detach().appendTo('.header-inner');
          var datas = isXml ? $("entry", res).map(function() {
            return {
              title: $("title", this).text(),
              content: $("content",this).text(),
              url: $("url" , this).text()
            };
          }).get() : res;
          var input = document.getElementById(search_id);
          var resultContent = document.getElementById(content_id);
          var inputEventFunction = function() {
            var searchText = input.value.trim().toLowerCase();
            var keywords = searchText.split(/[\s\-]+/);
            if (keywords.length > 1) {
              keywords.push(searchText);
            }
            var resultItems = [];
            if (searchText.length > 0) {
              // perform local searching
              datas.forEach(function(data) {
                var isMatch = false;
                var hitCount = 0;
                var searchTextCount = 0;
                var title = data.title.trim();
                var titleInLowerCase = title.toLowerCase();
                var content = data.content.trim().replace(/<[^>]+>/g,"");
                var contentInLowerCase = content.toLowerCase();
                var articleUrl = decodeURIComponent(data.url);
                var indexOfTitle = [];
                var indexOfContent = [];
                // only match articles with not empty titles
                if(title != '') {
                  keywords.forEach(function(keyword) {
                    function getIndexByWord(word, text, caseSensitive) {
                      var wordLen = word.length;
                      if (wordLen === 0) {
                        return [];
                      }
                      var startPosition = 0, position = [], index = [];
                      if (!caseSensitive) {
                        text = text.toLowerCase();
                        word = word.toLowerCase();
                      }
                      while ((position = text.indexOf(word, startPosition)) > -1) {
                        index.push({position: position, word: word});
                        startPosition = position + wordLen;
                      }
                      return index;
                    }

                    indexOfTitle = indexOfTitle.concat(getIndexByWord(keyword, titleInLowerCase, false));
                    indexOfContent = indexOfContent.concat(getIndexByWord(keyword, contentInLowerCase, false));
                  });
                  if (indexOfTitle.length > 0 || indexOfContent.length > 0) {
                    isMatch = true;
                    hitCount = indexOfTitle.length + indexOfContent.length;
                  }
                }

                // show search results

                if (isMatch) {
                  // sort index by position of keyword

                  [indexOfTitle, indexOfContent].forEach(function (index) {
                    index.sort(function (itemLeft, itemRight) {
                      if (itemRight.position !== itemLeft.position) {
                        return itemRight.position - itemLeft.position;
                      } else {
                        return itemLeft.word.length - itemRight.word.length;
                      }
                    });
                  });

                  // merge hits into slices

                  function mergeIntoSlice(text, start, end, index) {
                    var item = index[index.length - 1];
                    var position = item.position;
                    var word = item.word;
                    var hits = [];
                    var searchTextCountInSlice = 0;
                    while (position + word.length <= end && index.length != 0) {
                      if (word === searchText) {
                        searchTextCountInSlice++;
                      }
                      hits.push({position: position, length: word.length});
                      var wordEnd = position + word.length;

                      // move to next position of hit

                      index.pop();
                      while (index.length != 0) {
                        item = index[index.length - 1];
                        position = item.position;
                        word = item.word;
                        if (wordEnd > position) {
                          index.pop();
                        } else {
                          break;
                        }
                      }
                    }
                    searchTextCount += searchTextCountInSlice;
                    return {
                      hits: hits,
                      start: start,
                      end: end,
                      searchTextCount: searchTextCountInSlice
                    };
                  }

                  var slicesOfTitle = [];
                  if (indexOfTitle.length != 0) {
                    slicesOfTitle.push(mergeIntoSlice(title, 0, title.length, indexOfTitle));
                  }

                  var slicesOfContent = [];
                  while (indexOfContent.length != 0) {
                    var item = indexOfContent[indexOfContent.length - 1];
                    var position = item.position;
                    var word = item.word;
                    // cut out 100 characters
                    var start = position - 20;
                    var end = position + 80;
                    if(start < 0){
                      start = 0;
                    }
                    if (end < position + word.length) {
                      end = position + word.length;
                    }
                    if(end > content.length){
                      end = content.length;
                    }
                    slicesOfContent.push(mergeIntoSlice(content, start, end, indexOfContent));
                  }

                  // sort slices in content by search text's count and hits' count

                  slicesOfContent.sort(function (sliceLeft, sliceRight) {
                    if (sliceLeft.searchTextCount !== sliceRight.searchTextCount) {
                      return sliceRight.searchTextCount - sliceLeft.searchTextCount;
                    } else if (sliceLeft.hits.length !== sliceRight.hits.length) {
                      return sliceRight.hits.length - sliceLeft.hits.length;
                    } else {
                      return sliceLeft.start - sliceRight.start;
                    }
                  });

                  // select top N slices in content

                  var upperBound = parseInt('1');
                  if (upperBound >= 0) {
                    slicesOfContent = slicesOfContent.slice(0, upperBound);
                  }

                  // highlight title and content

                  function highlightKeyword(text, slice) {
                    var result = '';
                    var prevEnd = slice.start;
                    slice.hits.forEach(function (hit) {
                      result += text.substring(prevEnd, hit.position);
                      var end = hit.position + hit.length;
                      result += '<b class="search-keyword">' + text.substring(hit.position, end) + '</b>';
                      prevEnd = end;
                    });
                    result += text.substring(prevEnd, slice.end);
                    return result;
                  }

                  var resultItem = '';

                  if (slicesOfTitle.length != 0) {
                    resultItem += "<li><a href='" + articleUrl + "' class='search-result-title'>" + highlightKeyword(title, slicesOfTitle[0]) + "</a>";
                  } else {
                    resultItem += "<li><a href='" + articleUrl + "' class='search-result-title'>" + title + "</a>";
                  }

                  slicesOfContent.forEach(function (slice) {
                    resultItem += "<a href='" + articleUrl + "'>" +
                      "<p class=\"search-result\">" + highlightKeyword(content, slice) +
                      "...</p>" + "</a>";
                  });

                  resultItem += "</li>";
                  resultItems.push({
                    item: resultItem,
                    searchTextCount: searchTextCount,
                    hitCount: hitCount,
                    id: resultItems.length
                  });
                }
              })
            };
            if (keywords.length === 1 && keywords[0] === "") {
              resultContent.innerHTML = '<div id="no-result"><i class="fa fa-search fa-5x" /></div>'
            } else if (resultItems.length === 0) {
              resultContent.innerHTML = '<div id="no-result"><i class="fa fa-frown-o fa-5x" /></div>'
            } else {
              resultItems.sort(function (resultLeft, resultRight) {
                if (resultLeft.searchTextCount !== resultRight.searchTextCount) {
                  return resultRight.searchTextCount - resultLeft.searchTextCount;
                } else if (resultLeft.hitCount !== resultRight.hitCount) {
                  return resultRight.hitCount - resultLeft.hitCount;
                } else {
                  return resultRight.id - resultLeft.id;
                }
              });
              var searchResultList = '<ul class=\"search-result-list\">';
              resultItems.forEach(function (result) {
                searchResultList += result.item;
              })
              searchResultList += "</ul>";
              resultContent.innerHTML = searchResultList;
            }
          }

          if ('auto' === 'auto') {
            input.addEventListener('input', inputEventFunction);
          } else {
            $('.search-icon').click(inputEventFunction);
            input.addEventListener('keypress', function (event) {
              if (event.keyCode === 13) {
                inputEventFunction();
              }
            });
          }

          // remove loading animation
          $(".local-search-pop-overlay").remove();
          $('body').css('overflow', '');

          proceedsearch();
        }
      });
    }

    // handle and trigger popup window;
    $('.popup-trigger').click(function(e) {
      e.stopPropagation();
      if (isfetched === false) {
        searchFunc(path, 'local-search-input', 'local-search-result');
      } else {
        proceedsearch();
      };
    });

    $('.popup-btn-close').click(onPopupClose);
    $('.popup').click(function(e){
      e.stopPropagation();
    });
    $(document).on('keyup', function (event) {
      var shouldDismissSearchPopup = event.which === 27 &&
        $('.search-popup').is(':visible');
      if (shouldDismissSearchPopup) {
        onPopupClose();
      }
    });
  </script>





  

  
  <script src="https://cdn1.lncld.net/static/js/av-core-mini-0.6.4.js"></script>
  <script>AV.initialize("VczhXsUOQorVhmpiiHBEDGYW-gzGzoHsz", "r4UWWCgL44rY2AcvBAPEhHW1");</script>
  <script>
    function showTime(Counter) {
      var query = new AV.Query(Counter);
      var entries = [];
      var $visitors = $(".leancloud_visitors");

      $visitors.each(function () {
        entries.push( $(this).attr("id").trim() );
      });

      query.containedIn('url', entries);
      query.find()
        .done(function (results) {
          var COUNT_CONTAINER_REF = '.leancloud-visitors-count';

          if (results.length === 0) {
            $visitors.find(COUNT_CONTAINER_REF).text(0);
            return;
          }

          for (var i = 0; i < results.length; i++) {
            var item = results[i];
            var url = item.get('url');
            var time = item.get('time');
            var element = document.getElementById(url);

            $(element).find(COUNT_CONTAINER_REF).text(time);
          }
          for(var i = 0; i < entries.length; i++) {
            var url = entries[i];
            var element = document.getElementById(url);
            var countSpan = $(element).find(COUNT_CONTAINER_REF);
            if( countSpan.text() == '') {
              countSpan.text(0);
            }
          }
        })
        .fail(function (object, error) {
          console.log("Error: " + error.code + " " + error.message);
        });
    }

    function addCount(Counter) {
      var $visitors = $(".leancloud_visitors");
      var url = $visitors.attr('id').trim();
      var title = $visitors.attr('data-flag-title').trim();
      var query = new AV.Query(Counter);

      query.equalTo("url", url);
      query.find({
        success: function(results) {
          if (results.length > 0) {
            var counter = results[0];
            counter.fetchWhenSave(true);
            counter.increment("time");
            counter.save(null, {
              success: function(counter) {
                var $element = $(document.getElementById(url));
                $element.find('.leancloud-visitors-count').text(counter.get('time'));
              },
              error: function(counter, error) {
                console.log('Failed to save Visitor num, with error message: ' + error.message);
              }
            });
          } else {
            var newcounter = new Counter();
            /* Set ACL */
            var acl = new AV.ACL();
            acl.setPublicReadAccess(true);
            acl.setPublicWriteAccess(true);
            newcounter.setACL(acl);
            /* End Set ACL */
            newcounter.set("title", title);
            newcounter.set("url", url);
            newcounter.set("time", 1);
            newcounter.save(null, {
              success: function(newcounter) {
                var $element = $(document.getElementById(url));
                $element.find('.leancloud-visitors-count').text(newcounter.get('time'));
              },
              error: function(newcounter, error) {
                console.log('Failed to create');
              }
            });
          }
        },
        error: function(error) {
          console.log('Error:' + error.code + " " + error.message);
        }
      });
    }

    $(function() {
      var Counter = AV.Object.extend("Counter");
      if ($('.leancloud_visitors').length == 1) {
        addCount(Counter);
      } else if ($('.post-title-link').length > 1) {
        showTime(Counter);
      }
    });
  </script>



  
<script>
(function(){
    var bp = document.createElement('script');
    var curProtocol = window.location.protocol.split(':')[0];
    if (curProtocol === 'https') {
        bp.src = 'https://zz.bdstatic.com/linksubmit/push.js';        
    }
    else {
        bp.src = 'http://push.zhanzhang.baidu.com/push.js';
    }
    var s = document.getElementsByTagName("script")[0];
    s.parentNode.insertBefore(bp, s);
})();
</script>


  

  

  

  

</body>
</html>


